lib/Differentiator/ReverseModeVisitor.cpp

```diff
@@ -1857,15 +1861,82 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
-                                    noLoc)
-                     .get();
-    return StmtDiff(call);
+
+    Expr* call = nullptr;
+
+    if (FD->getReturnType()->isReferenceType()) {
+      DiffRequest calleeFnForwPassReq;
+      calleeFnForwPassReq.Function = FD;
+      calleeFnForwPassReq.Mode = DiffMode::reverse_mode_forward_pass;
+      calleeFnForwPassReq.BaseFunctionName = FD->getNameAsString();
+      calleeFnForwPassReq.VerboseDiags = true;
+      FunctionDecl* calleeFnForwPassFD =
+          plugin::ProcessDiffRequest(m_CladPlugin, calleeFnForwPassReq);
+
+      assert(calleeFnForwPassFD &&
+             "Clad failed to generate callee function forward pass function");
+
+      // FIXME: We are using the derivatives in forward pass here
+      // If `expr_dx()` is only meant to be used in reverse pass, 
+      // (for example, `clad::pop(...)` expression and a corresponding
+      // `clad::push(...)` in the forward pass), then this can result in
+      // incorrect derivative or crash at runtime. Ideally, we should have
+      // a separate routine to use derivative in the forward pass. 
+
+      // We cannot reuse the derivatives previously computed because
+      // they might contain 'clad::pop(..)` expression.
+      if (isa<CXXMemberCallExpr>(CE)) {
+        Expr* derivedBase = baseDiff.getExpr_dx();
+        if (isCladArrayType(derivedBase->getType()))
+          CallArgs.push_back(derivedBase);
+        else
+          CallArgs.push_back(
+              BuildOp(UnaryOperatorKind::UO_AddrOf, derivedBase, noLoc));
+      }
+
+      for (std::size_t i = 0, e = CE->getNumArgs(); i != e; ++i) {
+        const Expr* arg = CE->getArg(i);
+        const ParmVarDecl* PVD = FD->getParamDecl(i);
+        StmtDiff argDiff = Visit(arg);
+        if (argDiff.getExpr_dx() && PVD->getType()->isReferenceType()) {
+          Expr* derivedArg = argDiff.getExpr_dx();
+          if (isCladArrayType(derivedArg->getType()))
+            CallArgs.push_back(derivedArg);
+          else
+            CallArgs.push_back(
+                BuildOp(UnaryOperatorKind::UO_AddrOf, derivedArg, noLoc));
+        } else
+          CallArgs.push_back(m_Sema.ActOnCXXNullPtrLiteral(noLoc).get());
+      }
+      if (isa<CXXMemberCallExpr>(CE)) {
+        Expr* baseE = baseDiff.getExpr();
+        call = BuildCallExprToMemFn(
+            baseE, calleeFnForwPassFD->getName(), CallArgs, calleeFnForwPassFD);     
+      } else {
+        call = m_Sema
+                  .ActOnCallExpr(getCurrentScope(),
+                                  BuildDeclRef(calleeFnForwPassFD), noLoc,
+                                  CallArgs, noLoc)
+                  .get();
+      }
+      auto callRes = StoreAndRef(call);
+      auto resValue =
+          utils::BuildMemberExpr(m_Sema, getCurrentScope(), callRes, "value");
+      auto resAdjoint =
+          utils::BuildMemberExpr(m_Sema, getCurrentScope(), callRes, "adjoint");
+      return StmtDiff(resValue, resAdjoint, resAdjoint);
```

Comment:
```suggestion
      return StmtDiff(resValue, nullptr, resAdjoint);
```
Call expressions do not have a corresponding derivative that can be used explicitly in the reverse-sweep, 
as we insert the derivative call expression implicitly.

Reply:
You will also need to change, `ReverseModeVisitor::DifferentiateVarDecl`:

```cpp
if (isDerivativeOfRefType) {
        initDiff = Visit(VD->getInit());
        if (!initDiff.getExpr_dx()) {
          VDDerivedType =
              ComputeAdjointType(VD->getType().getNonReferenceType());
          isDerivativeOfRefType = false;
        }
        VDDerivedInit = getZeroInit(VDDerivedType);
      }
```
to:

```cpp
if (isDerivativeOfRefType) {
        initDiff = Visit(VD->getInit());
        if (!initDiff.getForwSweepExpr_dx()) { // This line changed!!
          VDDerivedType =
              ComputeAdjointType(VD->getType().getNonReferenceType());
          isDerivativeOfRefType = false;
        }
        VDDerivedInit = getZeroInit(VDDerivedType);
      }
```